client/src/main/java/io/pravega/client/segment/impl/ConditionalOutputStream.java

```diff
@@ -0,0 +1,46 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.client.segment.impl;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Defines an ConditionalOutputStream for a segment.
+ * Allows data to be conditionally appended to the end of the segment
+ */
+public interface ConditionalOutputStream extends AutoCloseable {
+
+    /**
+     * Returns the name of the segment associated to this output stream.
+     *
+     * @return The name of the segment associated to this output stream.
+     */
+    public String getSegmentName();
+    
+    /**
+     * Writes the provided data to the SegmentOutputStream. The data will be written only if the
+     * SegmentOutputStream's writeOffset is currently expectedOffset.
+     * 
+     * This is a synchronous and blocking method.
+     * 
+     * @param data The data to be added to the segment.
+     * @param expectedOffset The location in the segment that the data will be written at.
+     * @throws SegmentSealedException If the segment is closed for modifications.
+     */
+    public boolean write(ByteBuffer data, long expectedOffset) throws SegmentSealedException;
```

Comment:
This is somewhat of a subjective call but the rule is that exceptions should be unexpected.
If for example we were not able to write the data because the storage system failed then that is clearly an exception. Here it depends on your point of view. There are two ways this API could be used:
1. You have a single writer at a time and occasionally switch which process is doing the writing. Similar to what we do with BK ledgers on the server side. 
2. There are many processes writing at the same time and contention is normal. 

Both are possible, in the first case an exception is called for. I was thinking on the second case when I was writing this, so it uses a return value. Though the class could be used in either situation.

Reply:
Ok.